File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/readers/events/ConsentWithdrawnReader.kt
package com.snowplowanalytics.snowplow_flutter_tracker.readers.events

import com.snowplowanalytics.snowplow.event.ConsentDocument
import com.snowplowanalytics.snowplow.event.ConsentWithdrawn

class ConsentWithdrawnReader(val values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val all: Boolean by values
    val documentId: String by values
    val version: String by values
    val name: String? by valuesDefault
    val documentDescription: String? by valuesDefault
    val consentDocuments: List<Map<String, Any>>? by valuesDefault
    private val processedConsentDocuments: List<ConsentDocument>? by lazy {
        consentDocuments?.let { it.map { ConsentDocumentReader(it).toConsentDocument() } }
    }

    fun toConsentWithdrawn(): ConsentWithdrawn {
        val consentWithdrawn = ConsentWithdrawn(all, documentId, version)
        name?.let { consentWithdrawn.documentName(it) }
        documentDescription?.let { consentWithdrawn.documentDescription(it) }
        processedConsentDocuments?.let { consentWithdrawn.documents(it) }
        return consentWithdrawn
    }
}

File: ios/Classes/readers/events/ConsentWithdrawnReader.swift
//
//  ConsentWithdrawnReader.swift
//  snowplow_flutter_tracker
//
//  Created by Matus Tomlein on 04/01/2022.
//

import Foundation
import SnowplowTracker

struct ConsentWithdrawnReader: Decodable {
    let all: Bool
    let documentId: String
    let version: String
    let name: String?
    let documentDescription: String?
    let consentDocuments: [ConsentDocumentReader]?
}

extension ConsentWithdrawnReader {
    func toConsentWithdrawn() -> ConsentWithdrawn {
        let event = ConsentWithdrawn()
            .all(all)
            .documentId(documentId)
            .version(version)
        if let name = self.name { event.name(name) }
        if let description = self.documentDescription { event.documentDescription(description) }
        if let documents = self.consentDocuments {
            let jsons = documents.map { $0.toConsentDocument().getPayload() }
            event.documents(jsons)
        }
        return event
    }
}
